freqtrade/optimize/hyperopt.py

```diff
@@ -410,6 +410,34 @@ def prepare_hyperopt_data(self) -> None:
         # Store non-trimmed data - will be trimmed after signal generation.
         dump(preprocessed, self.data_pickle_file)
 
+    def get_asked_points(self, n_points: int) -> List[List[Any]]:
+        '''
+        Enforce points returned from `self.opt.ask` have not been already evaluated
+
+        Steps:
+        1. Try to get points using `self.opt.ask` first
+        2. Discard the points that have already been evaluated
+        3. Retry using `self.opt.ask` up to 3 times
+        4. If still some points are missing in respect to `n_points`, random sample some points
+        5. Repeat until at least `n_points` points in the `asked_non_tried` list
+        6. Return a list with length truncated at `n_points`
+        '''
+        i = 0
+        asked_non_tried: List[List[Any]] = []
+        while i < 100 and len(asked_non_tried) < n_points:
+            if i < 3:
+                asked = self.opt.ask(n_points=n_points)
+            else:
+                asked = self.opt.space.rvs(n_samples=n_points * 5)
+            asked_non_tried += [x for x in asked
+                                if x not in self.opt.Xi
+                                and x not in asked_non_tried]
+            i += 1
+        if asked_non_tried:
+            return asked_non_tried[:n_points]
```

Comment:
Wouldn't you have to check the length here, too? otherwise we risk getting fewer points than requested in some edge-case scenarios (where the space is almost exhausted).

with n_points=10 - i could in theory request 100 random points - but never get > n_points - so `asked_non_tried` would still be only 5 or so.

It's an edge-case for sure (probably one you never hit) - but i'm not entirely sure if it's handled correctly downstream, as until now, we assumed that `.ask()` get's me the amount of points i asked for.

Reply:
yes makes sense